nixos/modules/services/network-filesystems/moosefs.nix

```diff
@@ -219,28 +246,70 @@ in {
       (lib.optional cfg.client.enable pkgs.moosefs) ++
       (lib.optional cfg.master.enable initTool);
 
-    networking.firewall.allowedTCPPorts =
-      (lib.optionals cfg.master.openFirewall [ 9419 9420 9421 ]) ++
-      (lib.optional cfg.chunkserver.openFirewall 9422);
-
-    # Ensure storage directories exist
-    systemd.tmpfiles.rules =
-         lib.optional cfg.master.enable "d ${cfg.master.settings.DATA_PATH} 0700 ${mfsUser} ${mfsUser}"
-      ++ lib.optional cfg.metalogger.enable "d ${cfg.metalogger.settings.DATA_PATH} 0700 ${mfsUser} ${mfsUser}"
-      ++ lib.optional cfg.chunkserver.enable "d ${cfg.chunkserver.settings.DATA_PATH} 0700 ${mfsUser} ${mfsUser}";
-
-    # Service definitions
-    systemd.services.mfs-master = lib.mkIf cfg.master.enable
-    ( systemdService "master" {
-      TimeoutStartSec = 1800;
-      TimeoutStopSec = 1800;
-      Restart = "no";
-    } masterCfg );
-
-    systemd.services.mfs-metalogger = lib.mkIf cfg.metalogger.enable
-      ( systemdService "metalogger" { Restart = "on-abnormal"; } metaloggerCfg );
-
-    systemd.services.mfs-chunkserver = lib.mkIf cfg.chunkserver.enable
-      ( systemdService "chunkserver" { Restart = "on-abnormal"; } chunkserverCfg );
-    };
+    networking.firewall.allowedTCPPorts = lib.mkMerge [
+      (lib.optionals cfg.master.openFirewall [ 9419 9420 9421 ])
+      (lib.optional cfg.chunkserver.openFirewall 9422)
+      (lib.optional (cfg.cgiserver.enable && cfg.cgiserver.openFirewall) cfg.cgiserver.settings.PORT)
+    ];
+
+    systemd.tmpfiles.rules = [
+      # Master directories
+      (lib.optionalString cfg.master.enable
+        "d ${cfg.master.settings.DATA_PATH} 0700 ${mfsUser} ${mfsUser} -")
+
+      # Metalogger directories
+      (lib.optionalString cfg.metalogger.enable
+        "d ${cfg.metalogger.settings.DATA_PATH} 0700 ${mfsUser} ${mfsUser} -")
+
+      # Chunkserver directories
+      (lib.optionalString cfg.chunkserver.enable
+        "d ${cfg.chunkserver.settings.DATA_PATH} 0700 ${mfsUser} ${mfsUser} -")
+    ] ++ lib.optionals (cfg.chunkserver.enable && cfg.chunkserver.hdds != null)
+        (map (dir: "d ${dir} 0755 ${mfsUser} ${mfsUser} -") cfg.chunkserver.hdds);
+
+    systemd.services = lib.mkMerge [
+      (lib.mkIf cfg.master.enable {
+        mfs-master = systemdService "master" {
+          TimeoutStartSec = 1800;
+          TimeoutStopSec = 1800;
+          Restart = "on-failure";
+          User = mfsUser;
+          preStart = lib.mkIf cfg.master.autoInit "${initTool}/bin/mfsmaster-init";
```

Comment:
This ends up in the `serviceConfig` where it gets ignored.

We should also modify the test to run it with `autoInit`:
```
diff --git a/nixos/tests/moosefs.nix b/nixos/tests/moosefs.nix
index 0dc08748b828..440795959c71 100644
--- a/nixos/tests/moosefs.nix
+++ b/nixos/tests/moosefs.nix
@@ -9,6 +9,7 @@ let
     services.moosefs.master = {
       enable = true;
       openFirewall = true;
+      autoInit = true;
       exports = [
         "* / rw,alldirs,admin,maproot=0:0"
         "* . rw"
@@ -66,8 +67,6 @@ in {
     # prepare master server
     master.start()
     master.wait_for_unit("multi-user.target")
-    master.succeed("mfsmaster-init")
-    master.succeed("systemctl restart mfs-master")
     master.wait_for_unit("mfs-master.service")
 
     metalogger.wait_for_unit("mfs-metalogger.service")
```

Reply:
The options here go into `serviceConfig`, where preStart is ignored. The `preStart` script needs to go into `services.mfs-master.preStart` (instead of `services.mfs-master.serviceConfig.preStart`).